Python/optimizer.c

```diff
@@ -779,6 +800,76 @@ translate_bytecode_to_trace(
 #undef DPRINTF
 }
 
+static int
+remove_unneeded_uops(_PyUOpInstruction *trace, int trace_length)
+{
+    // Stage 1: Replace unneeded SAVE_IP uops with NOP.
+    // Note that we don't enter stubs, those SAVE_IPs are needed.
+    int last_save_ip = -1;
+    int last_instr = -1;
+    bool need_ip = true;
+    for (int pc = 0; pc < trace_length; pc++) {
+        int opcode = trace[pc].opcode;
+        if (opcode == SAVE_CURRENT_IP) {
+            // Special case: never remove preceding SAVE_IP
+            last_save_ip = -1;
+        }
+        else if (opcode == SAVE_IP) {
+            if (!need_ip && last_save_ip >= 0) {
+                trace[last_save_ip].opcode = NOP;
+            }
+            need_ip = false;
+            last_save_ip = pc;
+        }
+        else if (opcode == JUMP_TO_TOP || opcode == EXIT_TRACE) {
+            last_instr = pc + 1;
+            break;
+        }
+        else {
+            // If opcode metadata flags has ERROR or DEOPT, set need_up to true
+            if (_PyOpcode_opcode_metadata[opcode].flags & (HAS_ERROR_FLAG | HAS_DEOPT_FLAG)) {
+                need_ip = true;
+            }
+        }
+    }
```

Comment:
assert here that ``last_instr >= 0``?

Reply:
No, I can just initialize it to 0 though. If the trace is empty we shouldn't get here but the code here won't do anything.